seq dict

lindenb opened the pull request:
added a few functions to SAMSequenceDictionary

* md5() to get a MD5 checksum of the dictionary ( to quickly find a dictionary in a `Map<String,SAMSequenceDictionary> ` )
* toString
* countBases return the number of bases seen so far if we're at position `chrom:pos`
* fraction : same as prevous, but return the ratio with the reference length

nh13 commented on src/java/htsjdk/samtools/SAMSequenceDictionary.java:
```diff
@@ -163,8 +165,72 @@ public boolean equals(Object o) {
+     * @return a MD5 checksum for this dictionary or the empty string if it is empty
+     */
+    public String md5() {
+        if(isEmpty()) return "";
```
@lindenb the coding style and elsewhere needs to be consistent with htsjdk.  It makes it easier for folks to maintain and read the code.  In this case, ```if(``` -> ```if (```.  Thanks for updating the code in this pull requests and future ones too, as I really appreciate the contributions you are making.  I would recommend using intelliJ's or your favorite IDE's code re-format function.

nh13 commented on src/java/htsjdk/samtools/SAMSequenceDictionary.java:
```diff
@@ -163,8 +165,72 @@ public boolean equals(Object o) {
+    public String md5() {
+        if(isEmpty()) return "";
+        try {
+            MessageDigest md5 = MessageDigest.getInstance("MD5");
```
IntelliJ would show a warning here, since this variable could be final.

nh13 commented on src/java/htsjdk/samtools/SAMSequenceDictionary.java:
```diff
@@ -163,8 +165,72 @@ public boolean equals(Object o) {
+        try {
+            MessageDigest md5 = MessageDigest.getInstance("MD5");
+            md5.reset();
+            for(SAMSequenceRecord ssr : mSequences)  {
```
final

nh13 commented on src/java/htsjdk/samtools/SAMSequenceDictionary.java:
```diff
@@ -163,8 +165,72 @@ public boolean equals(Object o) {
+            MessageDigest md5 = MessageDigest.getInstance("MD5");
+            md5.reset();
+            for(SAMSequenceRecord ssr : mSequences)  {
+                if(ssr.getSequenceIndex()>0)  md5.update((byte)' ');
```
missing space and extra space.

nh13 commented on src/java/htsjdk/samtools/SAMSequenceDictionary.java:
```diff
@@ -163,8 +165,72 @@ public boolean equals(Object o) {
+                md5.update(ssr.getSequenceName().getBytes());
+                md5.update(String.valueOf(ssr.getSequenceLength()).getBytes());
+            }
+            String  hash = new BigInteger(1, md5.digest()).toString(16);
```
```String  hash``` -> ```String hash```.  I'll stop now but you get the point

nh13 commented on src/java/htsjdk/samtools/SAMSequenceDictionary.java:
```diff
@@ -163,8 +165,72 @@ public boolean equals(Object o) {
+            for(SAMSequenceRecord ssr : mSequences)  {
+                if(ssr.getSequenceIndex()>0)  md5.update((byte)' ');
+                md5.update(ssr.getSequenceName().getBytes());
+                md5.update(String.valueOf(ssr.getSequenceLength()).getBytes());
```
Perhaps we also use the ```MD5_TAG``` as well if it is available?  This would really help in the case that the sequence names and lengths are the same, but the bases differ.  It happens, seriously.

nh13 commented on src/java/htsjdk/samtools/SAMSequenceDictionary.java:
```diff
@@ -163,8 +165,72 @@ public boolean equals(Object o) {
         return true;
     }
 
+    /** return a MD5 sum for ths dictionary, the checksum is not re-computed each time this
```
I don't see how it is not re-computed each time.  In the case that it doesn't re-compute, it perhaps should when new sequences are set.

nh13 commented on src/java/htsjdk/samtools/SAMSequenceDictionary.java:
```diff
@@ -163,8 +165,72 @@ public boolean equals(Object o) {
+    
+    @Override
+    public String toString() {
+        return mSequences.toString();
```
So this would return the class name + the hashcode of the array?  Perhaps it is more useful to return the md5?  What's the use case for this?

nh13 commented on src/java/htsjdk/samtools/SAMSequenceDictionary.java:
```diff
@@ -163,8 +165,72 @@ public boolean equals(Object o) {
+            return -1L;
+        }
+        long countBases = 0L; /* number of bases seen so far */
+        for(int i = 0 ; i < sequenceIndex ; i++) countBases += getSequence(i).getSequenceLength();
```
probably better to use ```for (final SAMSequenceRecord sequence : mSequeces) ...```.

nh13 commented on src/java/htsjdk/samtools/SAMSequenceDictionary.java:
```diff
@@ -163,8 +165,72 @@ public boolean equals(Object o) {
+    }
+    
+    /** shortcut to <pre>countBases(getSequenceIndex(sequenceName), position);</pre> */
+    public long  countBases(String sequenceName,final int position)  {
```
perhaps ```getTotalBases``` instead?

nh13 commented on src/java/htsjdk/samtools/SAMSequenceDictionary.java:
```diff
@@ -163,8 +165,72 @@ public boolean equals(Object o) {
+        return countBases(getSequenceIndex(sequenceName), position);
+    }    
+    
+    /** 
```
I would be hesitant to include these utility functions here as there hasn't been a great need for them and they are relatively easy to compute by client code.

nh13 commented on src/java/htsjdk/samtools/SAMSequenceDictionary.java:
```diff
@@ -163,8 +165,72 @@ public boolean equals(Object o) {
+     * @returns the number of bases seen from the first sequence at position 0 to the sequenceIndex/position or -1 on error*/
+    public long countBases(final int sequenceIndex,final int position)  {
+        if (sequenceIndex < 0 || position <0 || sequenceIndex >= mSequences.size() ) {
+            return -1L;
```
I would rather throw an exception here, rather than relying on a return code.

lindenb commented:
@nh13  Thank you for taking the time to review this ! I'll make the changes later but if you think it's not worth it, do feel free to close this pull request.

While I'm here. Would you accept the following changes : 

* adding JAXB annotations (@XmlRootElement ... ) to SamSequenceRecord and SAMSequenceDictionary , so we can serialize things as XML/JSON (following the CA4GH schema ?)

* adding a method `void addAlias` to help resolving user's query with alternate contig names (chr1, 1, 01 ). Something like:

```java
public void addAlias (String baseName,String newName)
  {
  if(baseName.equals(newName)) return;
  SAMSequenceRecord ssr = getSequence(baseName);
  if( ssr == null ) throw new IllegalArgumentException("no sequence with that basename");
  if(  getSequence(newName) != null) throw new IllegalArgumentException("newName already used");
  this.mSequenceMap.put(newName,ssr);
  }
```

lindenb commented:
* I've updated my 'format' settings for eclipse (hoping it's ok...)
* in SamSequenceDictionary:
   * added getTotalBases (removed `fraction` as you suggested)
   * added addSequenceAlias : adds a new name to the dictionary map without adding a new sequence
   * added a toString verbose
   * added JaxbElement for XML serialization

* created  SamSequenceDictionaryTest to test SamSequenceDictionary XML serialization and aliases
* added JAXB annotations to AbstractSAMHeaderRecord, SamSequenceDictionary, SAMSequenceRecord
* in SAMSequenceRecord now using the predefined attributes instead of plain string

nh13 commented on src/java/htsjdk/samtools/SAMSequenceDictionary.java:
```diff
@@ -163,8 +176,124 @@ public boolean equals(Object o) {
+    public SAMSequenceRecord addSequenceAlias(final String originalName,
+            final String altName) {
+        if (originalName == null)
+            throw new IllegalArgumentException("original name cannot be null");
```
either put this one one line, or use ```{...}```

nh13 commented on src/java/htsjdk/samtools/SAMSequenceDictionary.java:
```diff
@@ -163,8 +176,124 @@ public boolean equals(Object o) {
         return true;
     }
 
+    /**
```
Probably a good idea to update the comment in ```setSequences``` to say that aliases will not be overwritten.

nh13 commented on src/java/htsjdk/samtools/SAMSequenceDictionary.java:
```diff
@@ -163,8 +176,124 @@ public boolean equals(Object o) {
+            throw new IllegalArgumentException("Alias " + altName
+                    + " was already set to " + altSeqRecord.getSequenceName());
+        }
+        mSequenceMap.put(altName, originalSeqRecord);
```
Oh no!  Now ```size()``` wont return the correct number of records!

lindenb replied:
I put a new key in `mSequenceMap` but size() gets its values from `mSequences` isn't it ?

```java
    public int size() {
        return mSequences.size();
    }
```

nh13 commented on src/java/htsjdk/samtools/SAMSequenceDictionary.java:
```diff
@@ -163,8 +176,124 @@ public boolean equals(Object o) {
+        try {
+            final MessageDigest md5 = MessageDigest.getInstance("MD5");
+            md5.reset();
+            for (final SAMSequenceRecord ssr : mSequences) {
```
better variable name than ```ssr``` please.

nh13 commented on src/java/htsjdk/samtools/SAMSequenceDictionary.java:
```diff
@@ -163,8 +176,124 @@ public boolean equals(Object o) {
+
+    @Override
+    public String toString() {
+        return "SAMSequenceDictionary:( sequences:"+ size()+
```
if we had an alias and one contig, the size would be two!

nh13 commented on src/java/htsjdk/samtools/SAMSequenceDictionary.java:
```diff
@@ -163,8 +176,124 @@ public boolean equals(Object o) {
+            throw new IllegalArgumentException("getTotalBases("+sequenceIndex+","+position);
+        }
+        long countBases = 0L; /* number of bases seen so far */
+        for (int i = 0; i < sequenceIndex; i++)
```
```{ ... }```

nh13 commented on src/java/htsjdk/samtools/SAMSequenceDictionary.java:
```diff
@@ -163,8 +176,124 @@ public boolean equals(Object o) {
+     * @returns the number of bases seen from the first sequence at position 0
+     *          to the sequenceIndex/position or -1 on error
+     */
+    public long getTotalBases(final int sequenceIndex, final int position) {
```
Is it really appropriate to put utility functions like this here?  Could folks not compute this themselves?  I don't want to dump a ton of utility functions into these library classes.  Can you make a case for its usefulness for others?

nh13 commented:
Minor comments, looking great.  Make sure to format the code similar to the rest of the library's code.  @lindenb

lindenb commented:
removed 'getTotalBases' at the end, fixed indent , added some comments.
' hope it's ok.

yfarjoun commented:
:+1: thanks!
